Allow using LDAP as a password connector

aoldershaw opened the pull request:
<!--
Hi there! Thanks for submitting a pull request to Concourse!

The title of your pull request will be used to generate the release notes.
Please provide a brief sentence that describes the PR, using the [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Examples: "Add feature to doohickey", "Fix panic during spline reticulation"

We will edit the title if needed so don't worry about getting it perfect!

To help us review your PR, please fill in the following information.
-->

[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## What does this PR accomplish?
<!--
Choose all that apply.
Also, mention the linked issue here.
This will magically close the issue once the PR is merged.
-->
Bug Fix | **Feature** | Documentation

closes #6627.

## Changes proposed by this PR:
<!--
Tell the reviewer What changed, Why, and How were you able to accomplish that?
-->
* Add flag for configuring password connector (either ldap or local)
* Prohibit adding local users when password connector is ldap
* Update username/password login page style
* Allow configuring username prompt for ldap (appears on username/password login page)

## Notes to reviewer:
<!--
Leave a message to whoever is going to review this PR.
Mainly, pointers to review the PR, and how they can test it.
-->

Branched off of #6669 

## Release Note
<!--
If needed, you can leave a list of detailed descriptions here which will be
used to generate the release note for the next version of Concourse. The title
of the PR will also be pulled into the release note.

Example:
* Reticulating splines is the new process Concourse uses to create the network
  of lines between jobs.
* Combines many short lines and curves into a network of splines.
-->

* By setting `--password-connector` (`$CONCOURSE_PASSWORD_CONNECTOR`) to `ldap`, you can authenticate to Concourse with `fly login -u ... -p ...` using your LDAP credentials
  * Enabling this feature prohibits the use of local users
* If you use an attribute other than username for authenticating with LDAP (e.g. email address), you can now configure `--username-prompt` (`$CONCOURSE_USERNAME_PROMPT`) to change the help text when logging in via the UI

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [ ] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [ ] [Signed] all commits
- [ ] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

Commit:
B: allow configuring the password connector

currently the only supported connectors are local and ldap. the username
prompt for ldap can be configured - this appears on the login with
username/password screen (e.g. setting it to "Email" will replace
"Username" with "Email"). we had to hack around Dex's default username
prompt for the "local" connector, since it defaulted to "Email Address"

this commit also changes the style of the login with username/password
screen

Signed-off-by: Rui Yang <ryang@pivotal.io>
Co-authored-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw commented:
Marking as a draft because the styles for the login page aren't finalized

aoldershaw committed:
B: add autocomplete=off to password field

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
B: modify styles on login page

* spacing changes
* submit button has filled background
* cancel text is now an outlined red button

also, fixes the back navigation - previously, going back and selecting a
new connector always displayed "database error" - now, clicking cancel
takes you back to the login page directly, which generates a new state
token

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

aoldershaw committed:
hack: change ldap username prompt to email

Signed-off-by: Aidan Oldershaw <aoldershaw@pivotal.io>

Commit:
update dex web UI

Signed-off-by: Rui Yang <ruiya@vmware.com>

aoldershaw commented on skymarshal/dexserver/web/templates/password.html:
```diff
@@ -22,17 +22,25 @@ <h2 class="theme-heading">log in to your account</h2>
+      </div>
+      {{ if .BackLink }}
+      <div class="theme-form-row">
+        <button tabindex="3" id="cancel-back" type="button" class="dex-btn theme-btn-cancel" onclick="javascript:history.back()">cancel</button>
```
@xtremerui could you change the `onclick` listener to `onclick="location.href='/sky/login'"`? When you just go "back" to the previous page and try to select a new provider, it complains about a database error

xtremerui replied:
how did you produce that ? It seems working in my local. @aoldershaw

xtremerui replied:
hmm for some reason when i run it locally with `docker-compose -f docker-compose.yml -f hack/overrides/ldap.yml up` i see both LDAP and Local are password connectors, which should not be allowed.

xtremerui replied:
and yeah the history back seems not working with button click so Dex throws out an error that `connector ID is already set in the request`

chenbh replied:
1. We [generate an `auth_request` id](https://github.com/concourse/dex/blob/05bffc9535a22afaa3817e0a993c58668f2b2b9f/server/handlers.go#L150-L160) as soon as we navigate to `/sky/issuer/auth` 
2. It sets the `connector_id` as soon as you click one of the options and get navigated to `/sky/issuer/auth/ldap`
3. Clicking the back button puts us back at `/sky/issuer/auth`, but still retains the request id
4. Clicking the other provider (`/sky/issuer/auth/ldap`) will reuse the old request id, which fails because [the `connector_id` has already been set](https://github.com/concourse/dex/blob/05bffc9535a22afaa3817e0a993c58668f2b2b9f/server/handlers.go#L218-L237)

Is it possible for us to detect that the user clicked the back button, and force dex to generate a new `auth_request` id?

xtremerui replied:
@chenbh I think the way @aoldershaw suggested is forcing dex to regenerate a new `req`. 

When you switching between providers by clicking cancel button, the `req` in url `../sky/issuer/auth/ldap?req=vm3hkmgon5a33jejpww5gsd5j` is changing. Is that what you meant?

chenbh replied:
Yeah the "cancel" button is working as expected, I'm not sure how you would do make that work for the back button though

xtremerui replied:
@chenbh pushed a change to handle the history back event so the `req` could be reset

Commit:
fix cancel button onclick

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented:
@aoldershaw I have pushed the fix of the cancel button. But running it locally with LDAP integration that using auth code grant seems broken.

chenbh commented on skymarshal/dexserver/web/themes/concourse/styles.css:
```diff
@@ -96,44 +114,34 @@
-.theme-form-input:focus,
-.theme-form-input:active {
-  border-color: white;
+  font-size: 20px;
```
What was the reason for increasing the font size? The difference in font size between most other text (12px) and the login screen is pretty jarring.

xtremerui replied:
https://www.figma.com/file/fuoFSVcTGptiS21P2POA2XbN/Concourse-UI-kit-and-Experiments?node-id=2590%3A170427

it is based on the sketch. I am not sure 100% but it might want to align with the font size used in the button?

matthewpereira replied:
Sorry @xtremerui - that might be my fault. I just made a new example based on a screenshot, but the screenshot was made on a retina monitor, which might have skewed the font sizes.

It's probably better to reduce the font size back down to its original scale.

Commit:
add back button listener to reset param req to empty

so that the request id could be regenerated by dex
also roll back to history.back() for cancel button since now it will
be handled by the listener

Signed-off-by: Rui Yang <ruiya@vmware.com>

Commit:
make sure back button event triggering

goes back 2 pages to pop the 2 pushed state to ensure history
entries cleaness

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented on skymarshal/dexserver/web/templates/header.html:
```diff
@@ -5,6 +5,14 @@
+        window.history.pushState({}, '', url);
+        window.history.back();
+      };
+    </script>
```
seems like a push state is needed for ensuring the popstate event triggering.

Commit:
fix a config bug for ldap integration test files

Signed-off-by: Rui Yang <ruiya@vmware.com>

chenbh commented:
@xtremerui  It looks like resetting the `req` param doesn't actually do anything, the request id is part of the href for the buttons themselves (no matter how we manipulate the url, it'll always restore the same html elements). The only way to get new request ids is to reload the page.

I was able to get it working by reloading the page if the navigation type is from a back button:
```diff
diff --git a/skymarshal/dexserver/web/templates/login.html b/skymarshal/dexserver/web/templates/login.html
index 1ad7fc4a7..b038d04fc 100644
--- a/skymarshal/dexserver/web/templates/login.html
+++ b/skymarshal/dexserver/web/templates/login.html
@@ -1,6 +1,13 @@
 {{ template "header.html" . }}

 <div class="theme-panel">
+  <script type="text/javascript">
+    // if we navigated to this page via the back button, reload the page to
+    // regenerate the request ID
+    if(performance.navigation.type == 2){
+       location.reload();
+    }
+  </script>
   <h2 class="theme-heading">login to {{ issuer }} </h2>
   <div>
     {{ range $c := .Connectors }}
```

xtremerui commented:
@chenbh what browsering are you using? It seems working for me in Chrome.

chenbh commented:
Using Chrome, updating to latest 89.0.4389.114 did not do anything

https://user-images.githubusercontent.com/17577073/113879032-2494a580-9788-11eb-9ae1-3410286bc1be.mov

Commit:
detect back type of navigation to force reload

this method is now an experiment feature and doesn't support
safari and IE.
https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming/type

Signed-off-by: Rui Yang <ruiya@vmware.com>

chenbh reviewed (approved):
lgtm!

Just gonna run the font size by @matthewpereira, the left is the increased font size from this PR, the right from CI:

https://user-images.githubusercontent.com/17577073/114076754-08226700-9875-11eb-90ce-2b9b1eceb30c.mov

matthewpereira commented:
RE: Font size:

Is there any responsive quality to our display here? IIRC the original was sized so that it wouldn't break when you scaled the page down to something tiny (eg phone/tablet), but this new version looks wide enough that it might get clipped weirdly if we're not accounting for screen size.

Commit:
revert font size and layout changes

Signed-off-by: Rui Yang <ruiya@vmware.com>

xtremerui commented:
the styles are updated:
![Screen Shot 2021-04-12 at 5 09 25 PM](https://user-images.githubusercontent.com/1585949/114463071-efd28500-9bb1-11eb-9ce6-7cb72318bbd6.png)
